thewalrus/_hermite_multidimensional.py

```diff
@@ -143,4 +148,160 @@ def hafnian_batched(A, cutoff, mu=None, rtol=1e-05, atol=1e-08, renorm=False, ma
+    """
+    copy = tup[:]
+    return tuple_setitem(copy, i, tup[i] - 1)
+
+
+@jit(nopython=True)
+def remove(
+    pattern: Tuple[int, ...]
+) -> Generator[Tuple[int, Tuple[int, ...]], None, None]:  # pragma: no cover
+    r"""returns a generator for all the possible ways to decrease elements of the given tuple by 1
+
+    Args:
+        pattern (Tuple[int, ...]): the pattern given to be decreased
+
+    Returns:
+        Generator[Tuple[int, Tuple[int, ...]], None, None]: the generator
+    """
+    for p, n in enumerate(pattern):
+        if n > 0:
+            yield p, dec(pattern, p)
+
+
+SQRT = np.sqrt(np.arange(1000))  # saving the time to recompute square roots
+
+
+def hermite_multidimensional_numba(R, cutoff, y, C = 1, dtype=np.complex128):
+    # pylint: disable=too-many-arguments
+    r"""Returns the multidimensional Hermite polynomials :math:`C*H_k^{(R)}(y)`.
+
+    Here :math:`R` is an :math:`n \times n` square matrix, and
+    :math:`y` is an :math:`n` dimensional vector. The polynomials are
+    parametrized by the multi-index :math:`k=(k_0,k_1,\ldots,k_{n-1})`,
+    and are calculated for all values :math:`0 \leq k_j < \text{cutoff}`,
+
+    Args:
+        R (array[complex]): square matrix parametrizing the Hermite polynomial
+        cutoff (int): maximum size of the subindices in the Hermite polynomial
+        y (vector[complex]): vector argument of the Hermite polynomial
+        C (complex): first value of the Hermite polynomials, the default value is 1
+        dtype (data type): Specifies the data type used for the calculation
+
+    Returns:
+        array[data type]: the multidimensional Hermite polynomials
+    """
+    n, _ = R.shape
+    if y.shape[0] != n:
+        raise ValueError("The matrix R and vector y have incompatible dimensions")
+    num_modes = len(y) // 2
+    array = np.zeros(((cutoff,) * (2 * num_modes)), dtype=dtype)
+    array[(0,) * (2 * num_modes)] = C
+    all_idx = product(range(cutoff), repeat=2 * num_modes)
+    next(all_idx)
+    for idx in all_idx:
+        array = fill_hermite_multidimensional_numba_loop(array, idx, R, y)
+    return array
+
+
+@jit(nopython=True)
+def fill_hermite_multidimensional_numba_loop(array, idx, R, y):  # pragma: no cover
+    r"""Calculates the hermit multidimensional polynomial for a given index.
+
+    Args:
+        array (array[data type]): the multidimensional Hermite polynomials
+        idx (tuple): index of the gradients to be filled
+        R (array[complex]): square matrix parametrizing the Hermite polynomial
+        y (vector[complex]): vector argument of the Hermite polynomial
+
+    Returns:
+        array[data type]: the hermit multidimensional polynomial for a given index
+    """
+    i = 0
+    for i, val in enumerate(idx):
+        if val > 0:
+            break
+    ki = dec(idx, i)
+    u = y[i] * array[ki]
+    for l, kl in remove(ki):
+        u -= SQRT[ki[l]] * R[i, l] * array[kl]
```

Comment:
Here we have a minus sign because of the convention of writing the second degree part of the polynomial at the exponential with a minus sign. Is this OK, or shall we leave + signs everywhere to avoid confusion?